### PR TITLE
Audits `XrpTransaction` Protobuf Wrapper

### PR DESCRIPTION
**Pull request by mvadari**
## High Level Overview of Change

This PR adds additional logic validations and descriptive error throwing to the class `XrpTransaction`.

### Context of Change

Another component of our attempt to make protobuf wrapper classes as safe and informative as possible. An SDK safety net.

### Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] Tests (You added tests for code that already exists, or your new feature included in this PR)

## Before / After

`XrpTransaction` performs logic/error checking, based on the business logic in the docs.

## Test Plan

New tests added to confirm that errors are correctly thrown. Existing tests were edited to properly account for the new errors. CI passes.

## Future Tasks
This is part of a series, with all subtasks listed here: https://app.asana.com/0/1188384817180008/1193021408531788/f

### CODE REVIEW

File: src/XRP/model/xrp-transaction.ts [hunk 1]
```diff
@@ -67,39 +101,66 @@ export default class XrpTransaction {
...
 
-    const sourceTag = transaction.getSourceTag()?.getValue()
+    const signingPublicKey = transaction.getSigningPublicKey()?.getValue_asB64()
+    if (signingPublicKey == undefined) {
+      throw new XrpError(
+        XrpErrorType.MalformedProtobuf,
+        'Transaction protobuf is missing valid `signingPublicKey` field.',
+      )
+    }
+    if (!signers && signingPublicKey == '') {
```
Comment (tedkalaw) [inline src/XRP/model/xrp-transaction.ts:111]: `===`

File: src/XRP/model/xrp-transaction.ts [hunk 2]
```diff
@@ -67,39 +101,66 @@ export default class XrpTransaction {
         .map((signer) => XrpSigner.from(signer, xrplNetwork))
     }
 
-    const sourceTag = transaction.getSourceTag()?.getValue()
+    const signingPublicKey = transaction.getSigningPublicKey()?.getValue_asB64()
+    if (signingPublicKey == undefined) {
```
Comment (tedkalaw) [inline src/XRP/model/xrp-transaction.ts:105]: `===`

File: src/XRP/model/xrp-transaction.ts [hunk 3]
```diff
@@ -26,26 +27,59 @@ export default class XrpTransaction {
...
-    const signingPublicKey = transaction.getSigningPublicKey()?.getValue_asU8()
+    const fee = transaction.getFee()?.getDrops()
+    if (!fee) {
+      throw new XrpError(
+        XrpErrorType.MalformedProtobuf,
+        'Transaction protobuf is missing valid `fee` field.',
+      )
+    }
+    const intFee = Number(fee)
+    if (isNaN(intFee)) {
```
Comment (tedkalaw) [inline src/XRP/model/xrp-transaction.ts:69]: is `getDrops` a strong? i think we want to use `BigInteger` here instead cc @amiecorso

File: src/XRP/default-xrp-client.ts [hunk 4]
```diff
@@ -322,11 +322,7 @@ export default class DefaultXrpClient implements XrpClientDecorator {
             getTransactionResponse,
             this.network,
           )
-          if (!xrpTransaction) {
-            throw XrpError.paymentConversionFailure
-          } else {
-            payments.push(xrpTransaction)
-          }
+          payments.push(xrpTransaction)
```
Comment (tedkalaw) [inline src/XRP/default-xrp-client.ts:325]: is this change related? it seems reasonable but can't tell

**Review by tedkalaw — CHANGES_REQUESTED** [review]

File: src/XRP/model/xrp-transaction.ts (hunk 3 above)
Comment (mvadari) [reply src/XRP/model/xrp-transaction.ts:69]: I was leaving it stored as a `string` for now (`intFee` isn't used at all after this check), just using it as a temp variable to check if `fee` represents an integer value or not.

File: src/XRP/default-xrp-client.ts (hunk 4 above)
Comment (mvadari) [reply src/XRP/default-xrp-client.ts:325]: It's a result of `XrpTransaction.from()` not being able to return undefined anymore.

File: src/XRP/model/xrp-transaction.ts (hunk 2 above)
Comment (amiecorso) [reply src/XRP/model/xrp-transaction.ts:105]: https://howtodoinjava.com/typescript/equals-vs-strict-equals/  interesting, don't think I knew this.  What's the risk of using `==` specifically in the context of a comparison with `undefined`?  will it give erroneous results?  I'm wondering if there are other places where we need to update this.

Also, is there a typescript operator that compares not just for same type and same value, but same object?  kinda the way `===` essentially compares memory addresses in... ? java I think?

File: src/XRP/model/xrp-transaction.ts [hunk 5]
```diff
@@ -67,39 +101,66 @@ export default class XrpTransaction {
...
-        }
-        type = payment && XrpTransactionType.Payment
+        paymentFields = XrpPayment.from(payment, xrplNetwork)
+        type = XrpTransactionType.Payment
         break
       }
+      case Transaction.TransactionDataCase.TRANSACTION_DATA_NOT_SET: {
+        throw new XrpError(
+          XrpErrorType.MalformedProtobuf,
+          'Transaction protobuf is missing transaction data.',
```
Comment (amiecorso) [inline src/XRP/model/xrp-transaction.ts:146]:
```suggestion
          'Transaction protobuf is missing transaction data case.',
```

File: src/XRP/model/xrp-transaction.ts [hunk 6]
```diff
@@ -67,39 +101,66 @@ export default class XrpTransaction {
...
+        )
+      }
       default:
-        // Unsupported transaction type.
-        return undefined
+        throw new XrpError(
+          XrpErrorType.MalformedProtobuf,
+          'Transaction protobuf has an unsupported transaction type.',
```
Comment (amiecorso) [inline src/XRP/model/xrp-transaction.ts:152]:
```suggestion
          'Transaction protobuf is of a transaction type unsupported by xpring-js.',
```

File: src/XRP/xrp-client-decorator.ts [hunk 7]
```diff
@@ -115,7 +115,7 @@ export default interface XrpClientDecorator {
    * @throws An error if the transaction hash was invalid.
    * @returns An {@link XrpTransaction} object representing an XRP Ledger transaction.
    */
-  getPayment(transactionHash: string): Promise<XrpTransaction | undefined>
```
Comment (amiecorso) [inline src/XRP/xrp-client-decorator.ts:118]: so satisfying

File: test/XRP/fakes/fake-xrp-protobufs.ts [hunk 8]
```diff
@@ -467,6 +500,127 @@ testInvalidPaymentTransaction.setSigningPublicKey(signingPublicKeyProto)
 testInvalidPaymentTransaction.setTransactionSignature(transactionSignatureProto)
 testInvalidPaymentTransaction.setPayment(testInvalidPaymentProtoNoAmount)
 
+// ... no account
```
Comment (amiecorso) [inline test/XRP/fakes/fake-xrp-protobufs.ts:503]: extremely great job on all these invalid test cases.  I'm going to be annoying and ask for one more:  can you make a fake transaction proto that's fine with the exception of a bad `Payment` as its `transaction` field?  I know we have a bad payment protobuf just laying around so hopefully it's simple enough.  It'll get us codecov on the one missing line.

File: test/XRP/xrp-protocol-buffer-conversion.test.ts [hunk 9]
```diff
@@ -651,13 +710,136 @@ describe('Protocol Buffer Conversion', function (): void {
...
+      XrpTransaction.from(
+        testInvalidGetTransactionResponseProtoNoSigningPublicKey,
+        XrplNetwork.Test,
+      )
+    }, XrpError)
+  })
+
+  it('Convert PAYMENT Transaction with no signers', function (): void {
+    // GIVEN a GetTransactionResponse protocol buffer with a Transaction field missing a signers field
+    // with an empty signingPublicKey field
```
Comment (amiecorso) [inline test/XRP/xrp-protocol-buffer-conversion.test.ts:801]:
```suggestion
```

File: test/XRP/xrp-protocol-buffer-conversion.test.ts [hunk 10]
```diff
@@ -651,13 +710,136 @@ describe('Protocol Buffer Conversion', function (): void {
...
+      XrpTransaction.from(
+        testInvalidGetTransactionResponseProtoNoSignature,
+        XrplNetwork.Test,
+      )
+    }, XrpError)
+  })
+
+  it('Convert PAYMENT Transaction with no data', function (): void {
+    // GIVEN a GetTransactionResponse protocol buffer with a Transaction field missing a data field
+    // with an empty signingPublicKey field
```
Comment (amiecorso) [inline test/XRP/xrp-protocol-buffer-conversion.test.ts:824]:
```suggestion
```

File: test/XRP/xrp-protocol-buffer-conversion.test.ts [hunk 11]
```diff
@@ -651,13 +710,136 @@ describe('Protocol Buffer Conversion', function (): void {
...
+      XrpTransaction.from(
+        testInvalidGetTransactionResponseProtoNoData,
+        XrplNetwork.Test,
+      )
+    }, XrpError)
+  })
+
+  it('Convert PAYMENT Transaction with no hash', function (): void {
+    // GIVEN a GetTransactionResponse protocol buffer with a Transaction field missing a hash field
+    // with an empty signingPublicKey field
```
Comment (amiecorso) [inline test/XRP/xrp-protocol-buffer-conversion.test.ts:836]:
```suggestion
```

**Review by amiecorso — CHANGES_REQUESTED** [review]
lookin good! just a few requests

**Comment by amiecorso** [comment]
considering pinging us on slack when you've updated the PR because I've found that "re-request review" often fails to put the PR back in our queue...  really annoying

File: src/XRP/model/xrp-transaction.ts (hunk 3 above)
Comment (tedkalaw) [reply src/XRP/model/xrp-transaction.ts:69]: seems reasonable to me - just wanted to make sure there wasn't a "standard" way we validate amounts elsewhere. gogogo

File: src/XRP/model/xrp-transaction.ts (hunk 2 above)
Comment (tedkalaw) [reply src/XRP/model/xrp-transaction.ts:105]: > https://howtodoinjava.com/typescript/equals-vs-strict-equals/ interesting, don't think I knew this. What's the risk of using == specifically in the context of a comparison with undefined? will it give erroneous results? I'm wondering if there are other places where we need to update this.

in this case, probably nothing - the only thing that `== undefined` is `null`. however, `==` does have some interesting behaviors: https://www.ecma-international.org/ecma-262/5.1/#sec-11.9.3

as such it's pretty common to turn on linting (https://eslint.org/docs/rules/eqeqeq) and have only one equality operator

> Also, is there a typescript operator that compares not just for same type and same value, but same object? kinda the way === essentially compares memory addresses in... ? java I think?

`===` compares memory addresses!! this is really important to how libraries like `React` and other functional and functional-y libraries work, particularly where places where changes need to be listened for

```
const obj1 = {}
const obj2 = {}
obj1 === obj2
> false
const obj1Ref = obj1
obj1 === obj1Ref
> true

File: src/XRP/model/xrp-transaction.ts (hunk 5 above)
Comment (mvadari) [reply src/XRP/model/xrp-transaction.ts:146]: My understanding of how the protobuf object works is that the data case gets set automatically, based on which data it has. So if it has no data case, then it's because it has no data.

File: test/XRP/fakes/fake-xrp-protobufs.ts (hunk 8 above)
Comment (mvadari) [reply test/XRP/fakes/fake-xrp-protobufs.ts:503]: It won't, because a bad `Payment` will trigger an error in `XrpPayment`. The one missing line is unreachable, but I can't think of a way to get rid of it, because as far as Typescript is aware, `payment` can still be `undefined`, even though the protobuf object won't let that be true when the data case is `Payment`.

File: src/XRP/model/xrp-transaction.ts (hunk 3 above)
Comment (mvadari) [reply src/XRP/model/xrp-transaction.ts:69]: Haven't seen this sort of check being applied anywhere, so idk if there's a "standard" way

Comment (amiecorso) [reply src/XRP/model/xrp-transaction.ts:69]: yeah nope there's not

File: src/XRP/model/xrp-transaction.ts (hunk 5 above)
Comment (amiecorso) [reply src/XRP/model/xrp-transaction.ts:146]: I think you're right about that, but I added this word just to make it clear that it's not the "transaction" field that's missing, but rather the type... since this is mostly human-facing.  Either way works for me

File: test/XRP/fakes/fake-xrp-protobufs.ts (hunk 8 above)
Comment (amiecorso) [reply test/XRP/fakes/fake-xrp-protobufs.ts:503]: Ohhh I see.  I guess one move would be to remove that check entirely because it probably is an impossible case logically, but that seems less defensive and unnecessary.  We can take the codecov hit

**Review by amiecorso — APPROVED** [review]
lookin good to me

File: test/XRP/fakes/fake-xrp-protobufs.ts (hunk 8 above)
Comment (mvadari) [reply test/XRP/fakes/fake-xrp-protobufs.ts:503]: Unfortunately Typescript gets mad if it's removed, since it could be `undefined`, so I don't think it can be removed.

**Review by tedkalaw — APPROVED** [review]
